http/sys_rekey.go

```diff
@@ -231,23 +235,161 @@ func handleSysRekeyUpdate(core *vault.Core, recovery bool) http.Handler {
+			respondError(w, http.StatusBadRequest,
+				fmt.Errorf("rekeying can only be performed on the primary cluster when replication is activated"))
+			return
+		}
+
+		ctx, cancel := core.GetContext()
+		defer cancel()
+
+		switch {
+		case recovery && !core.SealAccess().RecoveryKeySupported():
+			respondError(w, http.StatusBadRequest, fmt.Errorf("recovery rekeying not supported"))
+		case r.Method == "GET":
+			handleSysRekeyVerifyGet(ctx, core, recovery, w, r)
+		case r.Method == "POST" || r.Method == "PUT":
+			handleSysRekeyVerifyPut(ctx, core, recovery, w, r)
+		case r.Method == "DELETE":
+			handleSysRekeyVerifyDelete(core, recovery, w, r)
+		default:
+			respondError(w, http.StatusMethodNotAllowed, nil)
+		}
+	})
+}
+
+func handleSysRekeyVerifyGet(ctx context.Context, core *vault.Core, recovery bool, w http.ResponseWriter, r *http.Request) {
+	barrierConfig, barrierConfErr := core.SealAccess().BarrierConfig(ctx)
+	if barrierConfErr != nil {
+		respondError(w, http.StatusInternalServerError, barrierConfErr)
+		return
+	}
+	if barrierConfig == nil {
+		respondError(w, http.StatusBadRequest, fmt.Errorf("server is not yet initialized"))
+		return
+	}
+
+	// Get the rekey configuration
+	rekeyConf, err := core.RekeyConfig(recovery)
+	if err != nil {
+		respondError(w, err.Code(), err)
+		return
+	}
+	if rekeyConf == nil {
+		respondError(w, http.StatusBadRequest, errors.New("no rekey configuration found"))
+		return
+	}
+
+	// Get the progress
+	progress, err := core.RekeyVerifyProgress(recovery)
+	if err != nil {
+		respondError(w, err.Code(), err)
+		return
+	}
+
+	// Format the status
+	status := &RekeyVerificationStatusResponse{
+		Nonce:    rekeyConf.VerificationNonce,
+		T:        rekeyConf.SecretThreshold,
+		N:        rekeyConf.SecretShares,
+		Progress: progress,
+	}
+	respondOk(w, status)
+}
+
+func handleSysRekeyVerifyDelete(core *vault.Core, recovery bool, w http.ResponseWriter, r *http.Request) {
+	err := core.RekeyVerifyRestart(recovery)
+	if err != nil {
+		respondError(w, err.Code(), err)
+		return
+	}
+
+	ctx, cancel := core.GetContext()
+	defer cancel()
+
+	handleSysRekeyVerifyGet(ctx, core, recovery, w, r)
+}
+
+func handleSysRekeyVerifyPut(ctx context.Context, core *vault.Core, recovery bool, w http.ResponseWriter, r *http.Request) {
+	// Parse the request
+	var req RekeyVerificationUpdateRequest
+	if err := parseRequest(r, w, &req); err != nil {
+		respondError(w, http.StatusBadRequest, err)
+		return
+	}
+	if req.Key == "" {
+		respondError(
+			w, http.StatusBadRequest,
+			errors.New("'key' must be specified in request body as JSON"))
+		return
+	}
+
+	// Decode the key, which is base64 or hex encoded
+	min, max := core.BarrierKeyLength()
+	key, err := hex.DecodeString(req.Key)
+	// We check min and max here to ensure that a string that is base64
+	// encoded but also valid hex will not be valid and we instead base64
+	// decode it
+	if err != nil || len(key) < min || len(key) > max {
```

Comment:
Are the length comparisons sufficient? e.g. 'A'*40 decodes as  30 bytes in b64, 20 bytes in hex, both within a single/double aes block length. But they're not blocksize multiples, so perhaps this is not an issue.

Reply:
This is existing code that has been working for years, so don't want to modify it now.